File: backend/automation/tag_csv.py
"""Comma-separated tag lists: parse tokens and normalize stored form."""

from __future__ import annotations


def _normalize_one_tag_segment(p: str) -> str:
    """Trim and collapse internal runs of whitespace in one CSV segment."""
    t = " ".join(str(p).split())
    return t


def parse_tag_tokens(s: str | None) -> list[str]:
    return [
        t
        for t in (_normalize_one_tag_segment(p) for p in str(s or "").split(","))
        if t
    ]


def normalize_tag_csv(s: str | None, *, max_len: int = 200) -> str:
    parts = parse_tag_tokens(s)
    if not parts:
        return ""
    while parts:
        out = ", ".join(parts)
        if len(out) <= max_len:
            return out
        if len(parts) == 1:
            return parts[0][:max_len]
        parts.pop()
    return ""

File: frontend/src/utils/tagCsv.js
/**
 * Trims and collapses internal whitespace in one tag (CSV segment).
 * @param {string} s
 * @returns {string}
 */
export function normalizeTagName(s) {
  return String(s ?? "")
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * Splits a comma-separated tag list and normalizes each tag.
 * @param {string|null|undefined} raw
 * @returns {string[]}
 */
export function parseTagCsv(raw) {
  if (raw == null) return [];
  if (!String(raw).trim()) return [];
  return String(raw)
    .split(",")
    .map((p) => normalizeTagName(p))
    .filter(Boolean);
}

/**
 * @param {string|null|undefined} s
 * @returns {string}
 */
export function normalizeTagCsv(s) {
  return parseTagCsv(s).join(", ");
}

/**
 * Comma-separated JIRA keys; trim each segment (OR match when filtering).
 * @param {string|null|undefined} raw
 * @returns {string[]}
 */
export function parseJiraKeyCsv(raw) {
  if (raw == null || !String(raw).trim()) return [];
  return String(raw)
    .split(",")
    .map((p) => p.trim())
    .filter(Boolean);
}

/**
 * @param {string|null|undefined} s
 * @returns {string}
 */
export function normalizeJiraKeyCsv(s) {
  return parseJiraKeyCsv(s).join(", ");
}
